File: src/FSharpLint.Worker/IFSharpLintWorker.fs
(*
    FSharpLint, a linter for F#.
    Copyright (C) 2014 Matthew Mcveigh

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*)

namespace FSharpLint.Worker

type Range =
    {
        StartLine: int
        StartColumn: int
        EndLine: int
        EndColumn: int
        FileName: string
    }

type Error =
    {
        /// Description of the error.
        Info: string

        Range: Range

        /// Entire input file, needed to display where in the file the error occurred.
        Input: string

        FormattedError: string
    }

type Progress =
    | Starting of string
    | ReachedEnd of string
    | Failed of string * System.Exception

type LintOptions =
    {
        /// Function that when returns true cancels the parsing of the project, useful for cancellation tokens etc.
        FinishEarly: System.Func<bool>

        /// Callback that's called at the start and end of parsing each file (or when a file fails to be parsed).
        Progress: System.Action<Progress>

        /// Callback that's called when a lint error is detected.
        ErrorReceived: System.Action<Error>
    }

type Result =
    | Success
    | Failure of string

type IFSharpLintWorker =
    abstract member RunLint : projectFile:string -> options:LintOptions -> Result

File: src/FSharpLint.Worker/IFSharpLintWorker.cs
/// FSharpLint, a linter for F#.
/// Copyright (C) 2014 Matthew Mcveigh
/// 
/// This program is free software: you can redistribute it and/or modify
/// it under the terms of the GNU General Public License as published by
/// the Free Software Foundation, either version 3 of the License, or
/// (at your option) any later version.
/// 
/// This program is distributed in the hope that it will be useful,
/// but WITHOUT ANY WARRANTY; without even the implied warranty of
/// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
/// GNU General Public License for more details.
/// 
/// You should have received a copy of the GNU General Public License
/// along with this program.  If not, see <http://www.gnu.org/licenses/>.

namespace FSharpLint.Worker
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.ServiceModel;
    using System.Text;
    using System.Threading.Tasks;

    /// <summary>
    /// Handles lint errors received while linting a project.
    /// </summary>
    /// <param name="error">Lint error.</param>
    public delegate void ErrorReceivedEventHandler(Error error);

    /// <summary>
    /// Handles the current progress of lint running on a project.
    /// </summary>
    /// <param name="progress">Linter progress.</param>
    public delegate void ReportProgressEventHandler(Progress progress);

    /// <summary>
    /// Worker interface to be used to implement proxy objects to run lint in separate app domains.
    /// </summary>
    public interface IFSharpLintWorker
    {
        /// <summary>
        /// Reports any lint errors found while running lint.
        /// </summary>
        event ErrorReceivedEventHandler ErrorReceived;

        /// <summary>
        /// Reports progress made while running lint so we know what file is currently being linted.
        /// </summary>
        event ReportProgressEventHandler ReportProgress;

        /// <summary>
        /// Lints a given project file.
        /// </summary>
        /// <param name="projectFile">Path to the project file to be linted.</param>
        /// <returns>Result indicating whether the lint ran successfully or failed.</returns>
        Result RunLint(string projectFile);
    }
}
